### ISSUE TITLE
Serious floating-point artifacts on CPP FlxTilemap & FlxSprite

### ISSUE THREAD

**Issue by cwkx** [HaxeFlixel/flixel#911]
![pic](https://f.cloud.github.com/assets/5144374/2305507/4590e95e-a25a-11e3-9b98-1af90c01ffee.png)

I made all transparent tiles in my tileMap bright green to highlight the issue. The boundaries of some tiles are showing pixel data from their adjacent neighbours.

(1) This only occurs on some resolutions. Try lots of different resolutions.
(2) This occurs infrequently; best way to "find" it is by moving slowly in x and y axis. Most likely some floating point issue.

Using latest version from github.
##

**Comment by JoeCreates** [comment]
Just wanted to check what the state of this is these days, seeing as it's been a while. Any news or leads?

**Comment by RawArkanis** [comment]
As far I know you need to use `FlxG.camera.pixelPerfectRender = false;` in your FlxState.

**Comment by Tiago-Ling** [comment]
Hey, i solved my problem with artifacts in cpp targets (windows & android) by using the following combination:
* FlxG.camera.pixelPerfectRender = false;
* FlxG.camera.antialiasing = true;
* FlxSprite.pixelPerfectPosition = false; (applied to every sprite)

At least for FlxSprites this fixed the artifact problem in my project (using release versions of Lime, OpenFL and Hxcpp and dev version of HaxeFlixel).

However i'm still suffering from the 'blur' when moving the camera - it is much stronger than in any other framework or game i've seen so far. When the camera is moving fast the objects appear to be 'shaking'. Solving this would make the visual part of any scrolling game made with the framework a hundred times better.

**Comment by gamedevsam** [comment]
Could you try disabling antialiasing and see if the camera blur seems
reduced? I hadn't heard about the camera blur before, hope you figure out
how to fix it...

**Comment by Tiago-Ling** [comment]
I'll try that and to record a video showing the problem (it is hard because of video encoding, frame skipping and other things that hide the problem).

**Comment by Beeblerox** [comment]
i can confirm "blur" issue, but it's not blur, it's more like camera is slightly jumping.

**Comment by Beeblerox** [comment]
and it looks like different objects on the camera scrolls a little bit differently (just by one pixel)

**Comment by Beeblerox** [comment]
i think it might be something with camera's scroll behavior of with sprite onscreen position calculations

**Comment by Tiago-Ling** [comment]
If it has something to do with the camera, it seems to be down to two methods that might be part of the problem:
* updateFollow
* and/or updateScroll

Also FlxObject.updateMotion

**Comment by MSGhero** [comment]
My guess is `followLerp` since it was never definitively resolved how it should work with respect to the framerate.

**Comment by JoeCreates** [comment]
The main issue does seem to be a discrepancy between whatever position the texture rendering is trying to use and the actual rendered position of sprites. I wouldn't say this is particularly a camera issue (though the jerkiness may be a separate issue).

**Comment by Tiago-Ling** [comment]
By the way, disabling anti-aliasing seems to increase jerkiness / camera "blur".

After looking through the classes again i still have no solid opinion on what the problem might be.

I tried reducing my game's fps to 30 and the problem becomes far worse / more visible: the rendering / position mismatches become bigger.

I think these problems are not performance-related, but in any case, that's the screenshot of my specs running cpp in release mode (windows):
![perf_7_3_full](https://cloud.githubusercontent.com/assets/1617118/6220595/4928f6ae-b61f-11e4-9d7d-3d67e6597476.png)

**Comment by MSGhero** [comment]
@Tiago-Ling try messing with `updateFollow()`, specifically that `followLerp` part where it's like `60 / FlxG.framerate` or whatever. See if that affects your jerkiness.

**Comment by FryDay** [comment]
Any update on this? I have the same issue, but it only seems to happen with horizontal camera movement, not vertical. I am using TileMap and the auto-tile option. Something else that is interesting is that it seems to happen in fullscreen, but not windowed.

**Comment by camiloei** [comment]
i have the same problem with horizontal movements. 
I am using FlxTilemap to render my level

**Comment by FryDay** [comment]
@camiloei, out of curiosity, what OS are you using? I ask since I do all of my dev in Linux, which can sometimes introduce...other issues ;)

**Comment by camiloei** [comment]
@FryDay i'm using windows 7 64-bits, i only have this issue on neko and windows platforms (mobile not tested and flash works good)

**Comment by FryDay** [comment]
Gotcha. I just wanted to make sure it wasn't isolated to Linux. I get it on Linux and neko builds.

**Comment by luizdepra** [comment]
This is a problem with hxcpp based compilations. So, neko, windows, mac os and linux all have it.

**Comment by camiloei** [comment]
@luizdepra any solutions or workarounds?

**Comment by luizdepra** [comment]
@camiloei Nop. Tried @Tiago-Ling 's soluction but no sucess.

**Comment by Ohmnivore** [comment]
@luizdepra @camiloei If you scroll a good deal up i have posted a solution which worked perfectly for me. Like 100% perfect. I'll never understand why so many people continue to revive this "thread" when there is a solution already. Also i believe beeblerox included essentially the same solution in the dev branch. Don't quote me on this - i'm not sure but it's the impression i got.

In short it just adds a border and duplicates last rows and columns for every tile so that the GPU doesn't get confused when choosing the border pixels.

**Comment by Ohmnivore** [comment]
I think we also get 1-2 questions about this issue on the forums like every month lol

**Comment by Tiago-Ling** [comment]
What he says is true. By adding borders around the frames it's possible to
mitigate or even eliminate the problem.

This seems like a hardware "issue", in which sometimes the rect sent to the
gpu is 1px bigger.

In my experience, it can happen in frames with odd or non power of two
sizes.

Somebody else might know why this happens better, but the solution
@Ohmnivore describes works perfectly.

Also, some software like TexturePacker have an option to do this
automatically (look for " Extrusion" - thanks @Beeblerox)

On Sat, Sep 5, 2015, 12:02 AM Ohmnivore <notifications@github.com> wrote:

> I think we also get 1-2 questions about this issue on the forums like
> every month lol
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFlixel/flixel/issues/911#issuecomment-137896710>.
>
-- 
Tiago Ling Alexandre
Tel: +55 41 8819-3191

**Comment by camiloei** [comment]
@Tiago-Ling @Ohmnivore 

thank you guys! i will try your solutions.

**Comment by Beeblerox** [comment]
i think that we need to write article about it

**Comment by camiloei** [comment]
@Beeblerox i support this idea, it´s a very common problem and new people could have the same problem.

**Comment by MSGhero** [comment]
@Tiago-Ling if it's a hardware issue, can this be closed? We could write something up for the blog perhaps or sticky to the forums?

I'm procrastinating some work I have to do, btw. I wouldn't be this far back normally :smile:

**Comment by Gama11** [comment]
I don't think a blog or forum post is the right place - should be an article in the documentation.

**Comment by Tiago-Ling** [comment]
Should be included in the docs imho. Since this is a very common issue
people have it could be in a FAQ-style page in the site too?
This along with other super common stuff like correct camera usage could
help preventing too many beginners from going directly to github issues for
easily solvable problems.

**Comment by IMFUZZ** [comment]
I think I've found a pretty good solution. The problem seems to come from the _helperPoint which is not floored when calculating the drawX and drawY.

Using something like this completely fixes all the problems related to vertical lines appearing randomly when moving the camera around in my game.

var helperPointX = (Std.int(_helperPoint.x)%2 != 0) ? Math.floor(_helperPoint.x) : Std.int(_helperPoint.x);
var helperPointY = (Std.int(_helperPoint.y)%2 != 0) ? Math.floor(_helperPoint.y) : Std.int(_helperPoint.y);

I'ved encountered the problem when I tried to multiply a dark blue over the tilemap to fake a day/night cycle, blue lines were showing bright blue so they were easy to spot.

I'ved only tested the fix with 1280x720/640x360 resolution with 32x32 tiles and 16x16 tiles but the only time lines were showing was when resizing the window itself. Those lines disappears when activating and deactivating the pixelPerfectRender on the tilemap after resizing the windows to fullscreen.
